[vcpkg] Optionally limit the number of cores during vcpkg bootstrap

jherico opened the pull request:
Add an option to the bash bootstrap script that allows a user to force the use of only a single CPU during both compilation and linking of the vcpkg executable.  

Building on low-memory / high-core count systems like 1-2 GB variants of Raspberry PIs will often fail because of memory exhaustion.  Each compile will typically take a few hundred MB and running 4 of them concurrently will easily exhaust even a 2GB setup.  Further, such systems often have no swap space configured.  This command line option ensures that `vcpkg` will bootstrap successfully, albeit slowly, even when only 1GB of ram is available.

This PR _does not_ attempt to address any similar issues that might arise from building packages using `vcpkg`

- #### What does your PR fix?  

Fixes #14534 


- #### Which triplets are supported/not supported? Have you updated the [CI baseline]

This is a modification to the bootstrap script and should have no impact on the triplets, only a potential impact on the speed of bootstrapping.

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  

Yes.  The vast majority of the guide doesn't apply since this is a change to the boostrapping mechanism, not to any port or vcpkg functionality.  

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  

N/A

jherico committed:
limit the number of cores during vcpkg bootstrap

aizuon commented:
Could this feature also be implemented when compiling packages? Would be nice to leave half of the cores unattended when compiling big packages, i.e. Qt, to do other work to spend time.

jherico commented:
@aizuon It would be inconsistent.  As far as I understand, not every package uses CMake for the build wrapper, and the approach here only works with CMake.  It's likely that there may be some necessary similar feature that would be useful for package building, but this PR is just focused on the bootstrap process and unblocking people in the situation described above.

vicroms commented:
@aizuon You can try setting the [`VCPKG_MAX_CONCURRENCY`](https://vcpkg.io/en/docs/users/config-environment.html#vcpkg_max_concurrency) environment variable which most buildsystem helper functions support (there may be some ports with weird buildsystems that won't).

ras0219-msft commented:
I feel like it would make more sense to just make bootstrap respect `VCPKG_MAX_CONCURRENCY`

jherico committed:
Switch to VCPKG_MAX_CONCURRENCY

jherico commented:
@ras0219 I've modified the build script to respect `VCPKG_MAX_CONCURRENCY` and removed the other modifications.  I feel like it would be helpful to have some text in the `-help` output but I couldn't come up with something both informative and concise, so I left it as is.
